Proposal for more accurate player stats

ejfudd opened the pull request:
The player info panel now uses a calculated value for the stats instead of just the base values from the save file.  
The new calculation is as follows:
`Calculated stat = (base stat + bonuses from gear + bonuses from quests and masteries) * (total modifiers from gear, quests and masteries)`
It should now be a lot closer to the in game values though without accurate stat decoding for the items these cannot be 100% accurate.
The display red background for un-equippable items setting now takes into account the calculated player stat values as well as actually preventing equipping of those items.
The info panel is now more dynamic so equipping and un-equipping gear will change the displayed stat values.
The edit dialog still displays base stats.

Also added a couple of minor bug fixes:
The Allow character edit setting now will display/hide the button after closing the settings dialog and no longer requires loading in a new character or restarting TQVault.
The equipment panel now displays properly when no character is selected.  Previously, the medallion would be displayed partially over the top of the panel.

hguy commented on src/TQVaultAE.Data/ItemProvider.cs:
```diff
@@ -680,6 +707,30 @@ public bool FilterRequirements(string key)
+		/// <returns>True if the key is a stat boosting attrbute.</returns>
+		public bool IsStatBonus(string key)
+		{
+			string[] bonus =
```
putting that array in a static class field prevent unnecessary malloc using IsStatBonus()

ejfudd replied:
I will change it.  Should I change to other string arrays used for filtering as well?
The ones in FilterKey() and FilterRequirements()?

hguy replied:
IMHO it's better to have a "static readonly" for all these cases.

ejfudd replied:
Done.  Changed all of them to static readonly.

hguy commented:
The tooltip text "SettingsEnableCharacterRequierementBGColorTT" need adjustment to reflect these changes.

ejfudd commented:
Changed the TT to read

> When enabled, a red background color is displayed on items where the selected character does not meet the item requirements.
> The character will also not be able to equip items where the requirements are not met.

Unfortunately, I don't speak French so you will need to do that one.

hguy commented:
> Unfortunately, I don't speak French so you will need to provide that one.

Here it is
```Xml
<data name="SettingsEnableCharacterRequierementBGColorTT" xml:space="preserve">
    <value>Affiche une couleur de fond rouge sur les objets lorsque le personnage sélectionné ne répond pas aux exigences.
Vous ne serez égalament pas en mesure d'équiper l'objet sur ce personnage.
	</value>
  </data>
```

ejfudd commented:
Updated the French TT.
Is the label for the setting itself the same for English and French as it looks untranslated as well?

hguy commented:
Sorry there is a miss-typing
```
...
Vous ne serez également pas en mesure d'équiper l'objet sur ce personnage.
```

hguy commented:
> 
> 
> Updated the French TT.
> Is the label for the setting itself the same for English and French as it looks untranslated as well?

You right. Good eyes.
```Xml
  <data name="SettingsEnableCharacterRequierementBGColor" xml:space="preserve">
    <value>Active les restrictions liés aux exigences des objets</value>
  </data>
```

hguy commented:
All of this mean that "EnableCharacterRequierementBGColor" setting is not really a cosmetic control property.
It begin to control the overall tool logic by applying restriction.
Maybe we should rename it to avoid misconception.
Something like "EnableCharacterRequierementRestriction".

ejfudd commented:
Yes, I agree that renaming would make more sense.  I assume you mean changing the label in the Settings dialog as well as the variable names in the code. 
How about one of these for the label?
```
Enable Item Requirement Restriction
Enable Item Equip Requirements
Enable Item Requirements
```
The TT will provide more detailed info on what the setting actually does?

The variables can then be changed accordingly.

hguy commented:
The FR Label translation i provide already mean `Enables object requirements restrictions`. That's when i realized the shift in the logic.
I'm ok with your first suggestion or the shorter `Enable Requirement Restriction` as only items have requierements.

ejfudd commented:
Done. Updated the French strings.
Changed the English label on the Settings dialog.
Changed the variable names to match.
Changed the setting in the config file to match as well.

hguy commented on src/TQVaultAE.GUI/SettingsDialog.cs:
```diff
@@ -31,7 +31,7 @@ internal partial class SettingsDialog : VaultForm, IScalingControl
 		/// <summary>
 		/// Indicates that the EnableCharacterRequierementBGColor setting has been changed
```
Change comment with following so visual studio refactor and usage references can follow and automatically adjust it.
```CSharp
		/// Indicates that the <see cref="Config.Settings.EnableItemRequirementRestriction"/> setting has been changed

```

ejfudd replied:
Fixed.

hguy commented on src/TQVaultAE.GUI/SettingsDialog.cs:
```diff
@@ -156,7 +156,7 @@ internal partial class SettingsDialog : VaultForm, IScalingControl
 		/// <summary>
 		/// Indicates that the EnableCharacterRequierementBGColor setting has been changed
```
Same here
```CSharp
		/// Indicates that the <see cref="Config.Settings.EnableItemRequirementRestriction"/> setting has been changed

ejfudd replied:
Also fixed

hguy commented:
Beside my last comments, it looks good to me.
